### PR TITLE
Region annual reports

### PR DESCRIPTION
**Pull request by tetyanakutsko**

### CODE REVIEW

File: EPlast/EPlast.BLL/DTO/Region/RegionAnnualReportDTO.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+
+        public int NumberOfBeneficiaries { get; set; }
+
+
+        public int NumberOfPlastpryiatMembers { get; set; }
+
+        public int NumberOfHonoraryMembers { get; set; }
+
+        public User User { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.BLL/DTO/Region/RegionAnnualReportDTO.cs:61]: Why entity from DA exist in DTO ?

File: EPlast/EPlast.BLL/DTO/Region/RegionAnnualReportDTO.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+        public int NumberOfSeigneurMembers { get; set; }
+
+        public int NumberOfIndependentRiy { get; set; }
+
+        public int NumberOfClubs { get; set; }
+
+        public int NumberOfIndependentGroups { get; set; }
+
+
```
Comment (voroniak) [inline EPlast/EPlast.BLL/DTO/Region/RegionAnnualReportDTO.cs:44]: Remove unnecessary spaces

File: EPlast/EPlast.BLL/DTO/Region/RegionAnnualReportDTO.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        public int NumberOfTeachers { get; set; }
+
+
```
Comment (voroniak) [inline EPlast/EPlast.BLL/DTO/Region/RegionAnnualReportDTO.cs:47]: Remove unnecessary spaces

File: EPlast/EPlast.BLL/DTO/Region/RegionAnnualReportDTO.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        public int NumberOfAdministrators { get; set; }
+
+
```
Comment (voroniak) [inline EPlast/EPlast.BLL/DTO/Region/RegionAnnualReportDTO.cs:50]: Remove unnecessary spaces

File: EPlast/EPlast.BLL/Interfaces/Region/IRegionAccessGetter.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
+using DatabaseEntities = EPlast.DataAccess.Entities;
+
+namespace EPlast.BLL.Services.Region.RegionAccess.RegionAccessGetters
+{
+    public interface IRegionAccessGetter
+    {
+        Task<IEnumerable<DatabaseEntities.Region>> GetRegion(string userId);
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Interfaces/Region/IRegionAccessGetter.cs:9]: Why does the service returns entity from DA?

Comment (voroniak) [inline EPlast/EPlast.BLL/Interfaces/Region/IRegionAccessGetter.cs:9]:
```suggestion
        Task<IEnumerable<DatabaseEntities.Region>> GetRegionAsync(string userId);
```

File: EPlast/EPlast.BLL/Interfaces/Region/IRegionAnnualReportService.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        /// <summary>
+        /// Method to get region report by Id
+        /// </summary>
+        /// <param name="claimsPrincipal">Authorized user</param>
+        /// <param name="id">Annual report identification number</param>
+        /// <param name="year">Annual report identification number</param>
+        /// <returns>Annual report model</returns>
+        /// <exception cref="System.UnauthorizedAccessException">Thrown when user hasn't access to annual report</exception>
+        /// <exception cref="System.NullReferenceException">Thrown when annual report doesn't exist</exception>
+        Task<RegionAnnualReportDTO> GetReportById(int id,int year);
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Interfaces/Region/IRegionAnnualReportService.cs:47]:
```suggestion
        Task<RegionAnnualReportDTO> GetReportByIdAsync(int id,int year);
```

File: EPlast/EPlast.BLL/Interfaces/Region/IRegionAnnualReportService.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+
+        /// <summary>
+        /// Method to get all regions reports
+        /// </summary>
+        /// <returns>Annual report model</returns>
+        /// <exception cref="System.UnauthorizedAccessException">Thrown when user hasn't access to annual report</exception>
+        /// <exception cref="System.NullReferenceException">Thrown when annual report doesn't exist</exception>
+        Task<IEnumerable<RegionAnnualReportDTO>> GetAllRegionsReports();
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Interfaces/Region/IRegionAnnualReportService.cs:56]:
```suggestion
        Task<IEnumerable<RegionAnnualReportDTO>> GetAllRegionsReportsAsync();
```

File: EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+            return _mapper.Map<IEnumerable<RegionAnnualReport>, IEnumerable<RegionAnnualReportDTO>>(annualReports);
+        }
+
+        public async Task CreateAsync(ClaimsPrincipal claimsPrincipal, RegionAnnualReportDTO regionAnnualReportDTO)
+        {
+            var region = await _repositoryWrapper.Region.GetFirstOrDefaultAsync(
+                predicate: a => a.ID == regionAnnualReportDTO.Region.ID);
+            if (await CheckCreated(region.ID))
+            {
+                throw new InvalidOperationException();
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs:136]: It's better to add message to your exception

File: EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+            }
+            if (!await _regionAccessService.HasAccessAsync(claimsPrincipal, region.ID))
+            {
+                throw new UnauthorizedAccessException();
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs:140]: It's better to add message to your exception

File: EPlast/EPlast.DataAccess/Entities/Region/RegionAnnualReport.cs [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+        public int NumberOfHonoraryMembers { get; set; }
+
+        public MembersStatistic MembersStatistic { get; set; }
+
+        public int RegionId { get; set; }
+
+        public Region Region { get; set; }
+
+        public string RegionName { get; set; }
```
Comment (voroniak) [inline EPlast/EPlast.DataAccess/Entities/Region/RegionAnnualReport.cs:58]: As I understand Region contains Name, or not ?

File: EPlast/EPlast.WebApi/Controllers/RegionsController.cs [hunk 11]
```diff
@@ -122,10 +126,10 @@ public async Task<IActionResult> GetProfile(int regionId)
         [Authorize(AuthenticationSchemes = "Bearer", Roles = "Admin, Голова Округу")]
         public async Task<IActionResult> AddAdministrator(RegionAdministrationDTO admin)
         {
-                await _regionService.AddRegionAdministrator(admin);
+            await _regionService.AddRegionAdministrator(admin);
+
+            return Ok();
```
Comment (voroniak) [inline EPlast/EPlast.WebApi/Controllers/RegionsController.cs:131]: It's better to return other status code, I guess) like NoContent

**Review by voroniak — CHANGES_REQUESTED** [review]

File: EPlast/EPlast.BLL/DTO/Region/RegionAnnualReportDTO.cs (hunk 2 above)
Comment (tetyanakutsko) [reply EPlast/EPlast.BLL/DTO/Region/RegionAnnualReportDTO.cs:44]: Thanks. Removed.

File: EPlast/EPlast.BLL/DTO/Region/RegionAnnualReportDTO.cs (hunk 3 above)
Comment (tetyanakutsko) [reply EPlast/EPlast.BLL/DTO/Region/RegionAnnualReportDTO.cs:47]: Thanks. Removed.

File: EPlast/EPlast.BLL/DTO/Region/RegionAnnualReportDTO.cs (hunk 4 above)
Comment (tetyanakutsko) [reply EPlast/EPlast.BLL/DTO/Region/RegionAnnualReportDTO.cs:50]: Thanks. Removed.

File: EPlast/EPlast.BLL/DTO/Region/RegionAnnualReportDTO.cs (hunk 1 above)
Comment (tetyanakutsko) [reply EPlast/EPlast.BLL/DTO/Region/RegionAnnualReportDTO.cs:61]: Thanks. Removed.

File: EPlast/EPlast.BLL/Interfaces/Region/IRegionAccessGetter.cs (hunk 5 above)
Comment (tetyanakutsko) [reply EPlast/EPlast.BLL/Interfaces/Region/IRegionAccessGetter.cs:9]: Thanks. Changed.

File: EPlast/EPlast.BLL/Interfaces/Region/IRegionAnnualReportService.cs (hunk 6 above)
Comment (tetyanakutsko) [reply EPlast/EPlast.BLL/Interfaces/Region/IRegionAnnualReportService.cs:47]: Thanks. Changed.

File: EPlast/EPlast.BLL/Interfaces/Region/IRegionAnnualReportService.cs (hunk 7 above)
Comment (tetyanakutsko) [reply EPlast/EPlast.BLL/Interfaces/Region/IRegionAnnualReportService.cs:56]: Thanks. Changed.

File: EPlast/EPlast.DataAccess/Entities/Region/RegionAnnualReport.cs (hunk 10 above)
Comment (tetyanakutsko) [reply EPlast/EPlast.DataAccess/Entities/Region/RegionAnnualReport.cs:58]: Thanks. Changed.

File: EPlast/EPlast.WebApi/Controllers/RegionsController.cs (hunk 11 above)
Comment (tetyanakutsko) [reply EPlast/EPlast.WebApi/Controllers/RegionsController.cs:131]: done. Thanks

File: EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+            var regionAnnualReport = new RegionAnnualReport()
+            {
+                RegionName = region.RegionName,
+
+                RegionId = id,
+
+                Date = DateTime.Now,
+
+                NumberOfSeigneurMembers = region.Cities.Aggregate(0, (x, y) =>
+                           x += y.AnnualReports.Aggregate(0, (z, p) => z += p.MembersStatistic.NumberOfSeigneurMembers)),
```
Comment (LanchevychMaxym) [inline EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs:50]: Here you take statistic of all years, but I think you have to take only statistic of year that you pass in your method.

File: EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+                    x += y.AnnualReports.Aggregate(0, (z, p) => z += p.NumberOfHonoraryMembers)),
+
+                NumberOfAdministrators = region.Cities.Aggregate(0, (x, y) =>
+                    x += y.AnnualReports.Aggregate(0, (z, p) => z += p.NumberOfAdministrators)),
+
+                NumberOfTeacherAdministrators = region.Cities.Aggregate(0, (x, y) =>
+                    x += y.AnnualReports.Aggregate(0, (z, p) => z += p.NumberOfTeacherAdministrators)),
+
+                NumberOfTeachers = region.Cities.Aggregate(0, (x, y) =>
+                    x += y.AnnualReports.Aggregate(0, (z, p) => z += p.NumberOfTeachers)),
```
Comment (LanchevychMaxym) [inline EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs:110]: Here is the same issues I've described above

**Review by LanchevychMaxym — CHANGES_REQUESTED** [review]

File: EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs (hunk 8 above)
Comment (tetyanakutsko) [reply EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs:136]: Done. Thanks

File: EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs (hunk 12 above)
Comment (tetyanakutsko) [reply EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs:50]: Changed! Thanks

**Review by MykhailoKlapchuk — APPROVED** [review]

**Review by MykhailoKlapchuk — APPROVED** [review]

File: EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+            }
+            if (await CheckCreated(id,year))
+            {
+                throw new InvalidOperationException("Report is already crated!");
+            }
+            var cities = await _repositoryWrapper.City.GetAllAsync(
+                predicate: a => a.RegionId == id
+                );
+
+            var AnnualReports = new List<AnnualReport>();
```
Comment (LanchevychMaxym) [inline EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs:46]: Name AnnualReports does not match rule 'Local variables'. Suggested name is annualReports.

File: EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+            var MembersStatistics = new List<MembersStatistic>();
```
Comment (LanchevychMaxym) [inline EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs:47]: Name MembersStatisticsdoes not match rule 'Local variables'. Suggested name is membersStatistics.

**Review by LanchevychMaxym — CHANGES_REQUESTED** [review]

**Review by bodnarira — APPROVED** [review]

File: EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+
+            foreach(var city in cities)
+            {
+                var reports = await _repositoryWrapper.AnnualReports.GetAllAsync(a => a.Date.Year == year && a.CityId == city.ID);
+                if (reports.Count() > 0)
```
Comment (voroniak) [inline EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs:52]:
```suggestion
                if (reports.Count > 0)
```

File: EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs (hunk 15 above)
Comment (tetyanakutsko) [reply EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs]: Changed

File: EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs (hunk 14 above)
Comment (tetyanakutsko) [reply EPlast/EPlast.BLL/Services/Region/RegionAnnualReportService.cs]: Thanks! Changed

**Review by voroniak — APPROVED** [review]

**Review by LanchevychMaxym — APPROVED** [review]

**Pull request merged**